File: rabbit.tracking/src/main/java/rabbit/tracking/internal/util/Workbenches.java
/*
 * Copyright 2012 The Rabbit Eclipse Plug-in Project
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package rabbit.tracking.internal.util;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.collect.Sets.newHashSetWithExpectedSize;
import static java.util.Collections.unmodifiableSet;

import java.util.Set;

import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Shell;
import org.eclipse.ui.IPartService;
import org.eclipse.ui.IPerspectiveDescriptor;
import org.eclipse.ui.IWorkbench;
import org.eclipse.ui.IWorkbenchPage;
import org.eclipse.ui.IWorkbenchPart;
import org.eclipse.ui.IWorkbenchWindow;

public final class Workbenches {

  /**
   * Gets the current workbench window in focus.
   * 
   * @param workbench the workbench to get the focused window for
   * @return the currently focused window, or null if none
   * @throws NullPointerException if workbench is null
   */
  public static IWorkbenchWindow getFocusedWindow(final IWorkbench workbench) {
    check(workbench);

    if (Display.getCurrent() != null) {
      IWorkbenchWindow window = workbench.getActiveWorkbenchWindow();
      if (window == null) {
        return null;
      }
      return isFocused(window) ? window : null;
    }

    final IWorkbenchWindow[] result = {null};
    workbench.getDisplay().syncExec(new Runnable() {
      @Override public void run() {
        IWorkbenchWindow window = workbench.getActiveWorkbenchWindow();
        if (window != null && isFocused(window)) {
          result[0] = window;
        }
      }
    });
    return result[0];
  }

  /**
   * Gets the current workbench part in focus.
   * 
   * @param workbench the workbench to get the focused part for
   * @return the currently focused workbench part, or null if none
   * @throws NullPointerException if workbench is null
   */
  public static IWorkbenchPart getFocusedPart(IWorkbench workbench) {
    IWorkbenchWindow window = getFocusedWindow(check(workbench));
    return window != null ? window.getPartService().getActivePart() : null;
  }

  /**
   * Gets the current perspective in focus.
   * 
   * @param workbench the workbench to get the focused perspective for
   * @return the current focused perspective, or null if none
   * @throws NullPointerException if workbench is null
   */
  public static IPerspectiveDescriptor getFocusedPerspective(
      IWorkbench workbench) {
    IWorkbenchWindow window = getFocusedWindow(workbench);
    if (window == null) {
      return null;
    }
    IWorkbenchPage page = window.getActivePage();
    return page == null ? null : page.getPerspective();
  }

  /**
   * Gets all the {@link IPartService}s from the currently opened windows.
   * 
   * @param workbench the workbench to get all the part service for
   * @return a set of all {@link IPartService}s, not null, may be empty,
   *         unmodifiable
   * @throws NullPointerException if workbench is null
   */
  public static Set<IPartService> getPartServices(IWorkbench workbench) {
    IWorkbenchWindow[] windows = check(workbench).getWorkbenchWindows();
    Set<IPartService> services = newHashSetWithExpectedSize(windows.length);
    for (IWorkbenchWindow window : windows) {
      services.add(window.getPartService());
    }
    return unmodifiableSet(services);
  }

  private static boolean isFocused(IWorkbenchWindow window) {
    Shell shell = window.getShell();
    return shell.getDisplay().getActiveShell() == shell
        && !shell.getMinimized();
  }

  private static IWorkbench check(IWorkbench workbench) {
    return checkNotNull(workbench, "workbench");
  }

  private Workbenches() {
  }
}

File: rabbit.tracking.tests/src/main/scala/rabbit/tracking/tests/Workbenches.scala
/*
 * Copyright 2012 The Rabbit Eclipse Plug-in Project
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package rabbit.tracking.tests

import java.util.Random
import org.eclipse.ui.PlatformUI.getWorkbench
import org.eclipse.ui.{ IWorkbenchWindow, IWorkbenchPart, IViewPart }
import rabbit.tracking.tests.TestImplicits.funToRunnable
import org.eclipse.ui.IPerspectiveDescriptor
import java.util.concurrent.atomic.AtomicReference

object Workbenches {

  def activate(part: IWorkbenchPart) = {
    display.syncExec(() => {
      part.getSite.getPage.activate(part)
    })
    part
  }

  def close(window: IWorkbenchWindow) = display.syncExec(() => {
    if (window.getShell != null) {
      window.close()
    }
  })

  def closeAllParts(): Unit = display.syncExec(() => {
    closeAllParts(getWorkbench.getActiveWorkbenchWindow)
  })

  def closeAllParts(window: IWorkbenchWindow) = display.syncExec(() => {
    val page = window.getActivePage
    page.closeAllEditors(false)
    page.getViewReferences.foreach(page.hideView)
  })

  def closeAllPerspectives() { closeAllPerspectives(currentWindow) }

  def closeAllPerspectives(window: IWorkbenchWindow) = display.syncExec(() => {
    window.getActivePage.closeAllPerspectives(false, false)
  })

  def hide(view: IViewPart) = display.syncExec(() => {
    view.getSite.getPage.hideView(view)
  })

  def openRandomPerspective(): IPerspectiveDescriptor =
    openRandomPerspective(currentWindow)

  def openRandomPerspective(window: IWorkbenchWindow) = {
    var perspective = new AtomicReference[IPerspectiveDescriptor]()
    display.syncExec(() => {
      val current = window.getActivePage.getPerspective
      val currentId = if (current != null) current.getId else null
      val perspectives = window.getWorkbench.getPerspectiveRegistry.getPerspectives
      val page = window.getActivePage
      perspectives find (_.getId != currentId) match {
        case Some(_@ perspective) => page.setPerspective(perspective)
        case None => throw new RuntimeException("Don't have more perspective to show")
      }
      perspective set page.getPerspective
    })
    perspective.get
  }

  def openRandomPart(): IViewPart =
    openRandomPart(currentWindow)

  def openRandomPart(window: IWorkbenchWindow) = {
    var view = new AtomicReference[IViewPart]()
    display.syncExec(() => {
      val current = window.getActivePage.getActivePart
      val currentId = if (current != null) current.getSite.getId else null
      val views = window.getWorkbench.getViewRegistry.getViews
      val page = window.getActivePage
      views.find(_.getId != currentId) match {
        case Some(v) => view set window.getActivePage.showView(v.getId)
        case None => throw new RuntimeException("Don't have more view to show")
      }
    })
    view.get
  }

  def openWindow() = {
    var window = new AtomicReference[IWorkbenchWindow]()
    display.syncExec(() => {
      window set getWorkbench.openWorkbenchWindow(null)
      window.get.getShell.setActive
    })
    window.get
  }

  def currentWindow = {
    var window = new AtomicReference[IWorkbenchWindow]()
    display.syncExec(() => {
      window set getWorkbench.getActiveWorkbenchWindow
    })
    window.get
  }

  private def display = getWorkbench.getDisplay
}